Core/src/HttpMiddleware/RetryHeadersHttpMiddleware.php

```diff
@@ -0,0 +1,76 @@
+<?php
+/*
+ * Copyright 2023 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Core\HttpMiddleware;
+
+use Google\ApiCore\AgentHeader;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\ResponseInterface;
+use Ramsey\Uuid\Uuid;
+
+/**
+ * Add the retry headers (invocation ID and count) to the retried request
+ */
+class RetryHeadersHttpMiddleware
+{
+    private int $currentAttempt = 0;
+    private $nextHandler;
+
+    public function __construct($nextHandler)
+    {
+        $this->nextHandler = $nextHandler;
+    }
+
+    /**
+     * @param RequestInterface $request
+     * @param array $options
+     * @return ResponseInterface
+     */
+    public function __invoke(RequestInterface $request, array $options)
+    {
+        // increment the current attempt
+        $this->currentAttempt++;
```

Comment:
This logic could be in a method (say `work`) and we can put the work method in the interface for the middleware?

That way, the developer will put any application logic in the `work` method and the `_invoke` will have the boilerplate of calling `work` and the handler.

Reply:
Why not just typehint the invoke function?

```php
interface HttpMiddleware
{
    public function __invoke(RequestInterface $request, array $options);
}
```